pkg/chartutil/values.go

```diff
@@ -326,36 +275,50 @@ func coalesceValues(c *chart.Chart, v map[string]interface{}) (map[string]interf
 func coalesceTables(dst, src map[string]interface{}, chartName string) map[string]interface{} {
 	// Because dest has higher precedence than src, dest values override src
 	// values.
+
+	rv := make(map[string]interface{})
 	for key, val := range src {
 		dv, ok := dst[key]
-		if ok && dv == nil {
-			// skip here, we delete at end
+		if !ok { // if not in dst, then copy from src
+			rv[key] = val
 			continue
 		}
-		if istable(val) {
-			if !ok {
-				dst[key] = val
-			} else if istable(dv) {
-				coalesceTables(dv.(map[string]interface{}), val.(map[string]interface{}), chartName)
-			} else {
-				log.Printf("Warning: Merging destination map for chart '%s'. Cannot overwrite table item '%s', with non table value: %v", chartName, key, val)
-			}
+		if dv == nil { // if set to nil in dst, then ignore
+			// When the YAML value is null, we skip the value's key.
+			// This allows Helm's various sources of values (value files or --set) to
+			// remove incompatible keys from any previous chart, file, or set values.
 			continue
-		} else if ok && istable(dv) {
+		}
+
+		srcTable, srcIsTable := val.(map[string]interface{})
+		dstTable, dstIsTable := dv.(map[string]interface{})
+		switch {
+		case srcIsTable && dstIsTable: // both tables, we coalesce
+			rv[key] = coalesceTables(dstTable, srcTable, chartName)
+		case srcIsTable && !dstIsTable:
+			log.Printf("Warning: Merging destination map for chart '%s'. Cannot overwrite table item '%s', with non table value: %v", chartName, key, val)
+			// despite message in warning, we appear to do exactly that, and do take the dst value
```

Comment:
@scottrigby and I spent some time together going over the original algorithm and I am not sure how this comment is true. Based on the original logic there are 4 conditions (using `map` instead of "table" in all cases):

1. If the key does not exist in the destination map, set the incoming value
2. If the incoming value is a `map` and the existing key value is a `map`, merge
3. If the incoming value is not a `map` and the existing key value is a `map`, leave the existing value alone
4. If the existing key is not a `map` and the incoming value is a `map`, leave the existing value alone

Essentially 3 and 4 are the same and could be reworded as "if only one of the destination or incoming value is a `map`, leave the existing value alone"

The logic here reads much nicer (thanks!) but I think has different behavior. Were you observing something different in your testing? I also know that the original code is spaghetti code, so we could have read something wrong as well

Reply:
If we are correct (and please let us know if we are not!) I would clarify the log messages to be something of lines of "cannot override a non-map value with a map value" and remove the `rv[key] = dv` lines in those blocks

Reply:
@thomastaylor312 @scottrigby 

I added the following to `pkg/chartutil/values_test.go`:

```golang
func TestCannotOverriteTableItemWithNonTableValue(t *testing.T) {
	// src has a table value for "foo"
	src := map[string]interface{}{
		"foo": map[string]interface{}{
			"baz": "boz",
		},
	}

	// dst has a non-table value for "foo"
	dst := map[string]interface{}{
		"foo": "bar",
	}

	// warning claims that "foo" will not be overwritten, yet:
	log.Println("Result: ", coalesceTables(dst, src, ""))
}
```

Then:

```bash
$ go test ./pkg/chartutil -run TestCannotOverriteTableItemWithNonTableValue -v
=== RUN   TestCannotOverriteTableItemWithNonTableValue
2019/06/07 15:05:35 Warning: Merging destination map for chart ''. Cannot overwrite table item 'foo', with non table value: map[baz:boz]
2019/06/07 15:05:35 Result:  map[foo:bar]
--- PASS: TestCannotOverriteTableItemWithNonTableValue (0.00s)
PASS
ok      k8s.io/helm/pkg/chartutil       0.006s
```

The warning is wrong because:

1. The value that it prints is actually the `src` value, which is being overwritten *by* the `dst` value, although as written it implies that's the value that you are trying merge on top.
2. It's a no-op because it's the `dst` map (which has the `dst` value) that is actually returned, the `src` value is ignored in the result as shown in the output above.

The refactor preserves the above behavior and warning, even though I think the warning is inaccurate.

In terms of correct behavior - well I'm not entirely sure - and I haven't used Helm enough to get a good sense for it. We only got down this rabbit hole in trying to solve a bug that we encountered. My first impression is that the current behaviour, where the destination value totally overwrites the source in the event of a type mismatch is reasonable, however it shouldn't print the inaccurate warning. Alternatively one could take the view that a type mismatch should result in an error and fail processing. Either way, I think a warning is wrong. Either allow it or fail fast.

Reply:
Ok, that makes sense then, thanks for the clarification! And I agree with you about warning vs failing, but if we begin failing, it will break backwards compatibility (things will start breaking where they didn't before). Can you swap the warning message to be accurate and more clear? After that I think this should be good to go